text/0573-app-boot-hooks.md

```diff
@@ -0,0 +1,114 @@
+- Start Date: 2020-01-09
+- Relevant Team(s): Ember.js
+- RFC PR: https://github.com/emberjs/rfcs/pull/573
+- Tracking:
+
+# App Boot Hooks
+
+## Summary
+
+This RFC proposes adding two hooks in the `Application` class in `app.js` that, by default, load
+and run initializers and instance-initializers. Here's an eaxmple of what this could look like:
+
+```js
+import { loadInitializers, loadInstanceInitializers } from 'ember-load-initializers';
+import Application from '@ember/application';
+import config from './config/environment';
+
+export default class App extends Application {
+    onBoot() {
+        loadInitializers(this, config.modulePrefix);
+        this.runInitializers();
+    }
+    onInstanceBoot(appInstance) {
+        loadInstanceInitializers(this, config.modulePrefix)
+        this.runInstanceInitializers();
+    }
+}
+```
+
+This implementation would be backwards compatible and would give users complete control of
+*how* and *when* they want to run their initializers and it would allow the beginner experience to
+be progressively enhanced.
+
+## Motivation
+
+- **Moving away from initializers**. The existing convention of initializer code comes from
+Ruby on Rails (which has similar `config/initializers/*.rb` files). This technique works
+and can be nice for organizing code, but it also has several downsides:
+
+    - bookkeeping code to determine the order of execution
+    - confusing `before`/`after` API to change order of execution
+    - implicit injections from addons by way of merging directories
+    - lack of support for blocking asynchronous behavior for instance-initializers (`{defer|advance}Readiness` exists for initializers)
+    - a poor onboarding experience and misses out on a teaching opportunity about how Ember Applications work.
+
+    By providing a way to hook into the existing, async boot sequence more directly,
+    application developers can write declarative, explicit code about how they want to
+    customize their application.
+- **Blocking async code before routing**. Currently, the earliest place an application can write
+asynchronous blocking code is in the `beforeModel` hook of the `ApplicationRoute`.
+In some cases, this is too late. For example, if an application needs to wait for an event
+from its host environment e.g. an iframe or an electron shell, and use it to determine
+the starting URL, it cannot do so.
+- **First Transition**: Similarly, if one wants to add an event listener to the
```

Comment:
There's discussion first transition should be handled by an instance-initializer
https://github.com/emberjs/rfcs/pull/680#issuecomment-747739416

Will there be any recommendation change after this RFC?

Reply:
It can be handled by a function that runs in `boot()` instead. Note that this RFC doesn't fundamentally change the fact that you can run code during application boot. It just changes the developer ergonomics to enable new features.